src/javascript/search.js

```diff
@@ -0,0 +1,191 @@
+import API from './api.js';
+
+const api = new API();
+
+export default function Search({ keywordTarget, searchTarget, searchBarTarget, searchTextTarget, rollKeywordTarget, keywords }) {
+  this.keywordTarget = keywordTarget;
+  this.searchTarget = searchTarget;
+  this.searchBarTarget = searchBarTarget;
+  this.searchTextTarget = searchTextTarget;
+  this.rollKeywordTarget = rollKeywordTarget;
+  this.keywords = keywords;
+  this.timeOut = null;
+};
+
+Search.prototype = {
+  init,
+  renderKeywordList,
+  getKeywordBlock,
+  getTitleBlock,
+  createKeywordEvent,
+  clickKeyword,
+  createSearchTextEvent,
+  handleSearchTextEvent,
+  selectSearchTargetItem,
+  searchText,
+  getSearchList,
+  renderSearchList,
+  createSearchBarEvent,
+  leaveSearchBarEvent,
+  handleList,
+  focusInSearchText,
+  focusOutSearchText
+}
+
+Search.constructor = Search;
+
+function init() {
+  this.renderKeywordList();
+  this.createKeywordEvent();
+  this.createSearchTextEvent();
+  this.createSearchBarEvent();
+}
+
+function renderKeywordList() {
+  const list = this.keywords.reduce((acc, item, i) => acc += this.getKeywordBlock(item, i), '');
+  this.keywordTarget.insertAdjacentHTML('afterbegin', this.getTitleBlock());
+  this.keywordTarget.querySelector('ul').innerHTML = list;
+}
+
+function getKeywordBlock(item, i) {
+  return `<li data-value="${item}"><span class="keyword__rank">${i + 1}</span><span>${item}</span></li>`
+}
+
+function getTitleBlock() {
+  return '<p class="keyword__label">인기 쇼핑 키워드</p>';
+}
+
+function createKeywordEvent() {
+  this.keywordTarget.addEventListener('click', ({ target }) => this.clickKeyword(target));
+}
+
+function clickKeyword(target) {
+  if(target.tagName === 'SPAN') {
+    const value = target.closest('li').dataset.value;
+    this.searchTextTarget.value = value;
+    this.focusInSearchText();
+  }
+};
+
+function createSearchTextEvent() {
+  const { searchTextTarget } = this;
+  searchTextTarget.addEventListener('focusin', () => this.handleList('in'));
+  searchTextTarget.addEventListener('focusout', () => this.handleList('out'));
+  searchTextTarget.addEventListener('keydown', e => this.handleSearchTextEvent(e));
+}
+
+function handleSearchTextEvent(event) {
+  switch(event.key) {
+    case 'ENTER': break; // 로컬스토리지로 캐시 구현
+    case 'ArrowUp': this.selectSearchTargetItem(event, 'up'); break;
+    case 'ArrowDown': this.selectSearchTargetItem(event, 'down'); break;
+    default: this.searchText(event.target); break;
+  }
+}
+
+function searchText(target) {
+  if(this.timeOut) clearTimeout(this.timeOut);
+  this.timeOut = setTimeout(() => this.getSearchList(target.value), 1000);
+};
```

Comment:
`this.timeOut` 을 직접 제어하는 것 보다 여러 곳에서 재사용 가능한 debounce함수를 구현하고 내부에서 제어되도록 하는 것이 좋아 보여요 🙂
```suggestion
const searchText = debounce(function(target) { this.getSearchList(target.value); }, 1000);
```

Reply:
> `this.timeOut` 을 직접 제어하는 것 보다 여러 곳에서 재사용 가능한 debounce함수를 구현하고 내부에서 제어되도록 하는 것이 좋아 보여요 🙂

debounce 함수가 과제였는데도 미뤄놨던 부분에 뜨끔하네요...
전체적인 코드를 다시 한 번 검토해서 유틸 함수로 만들만한 부분을 분리하도록 하겠습니다 ☑️